kernel-laws/shared/src/test/scala/cats/kernel/laws/LawTests.scala

```diff
@@ -66,8 +66,20 @@ object KernelCheck {
   }
 
   // Copied from cats-laws.
-  implicit def arbitrarySortedSet[A: Arbitrary: Order]: Arbitrary[SortedSet[A]] =
-    Arbitrary(arbitrary[Set[A]].map(s => SortedSet.empty[A](implicitly[Order[A]].toOrdering) ++ s))
+  implicit def arbitrarySortedSet[A: Arbitrary: Order]: Arbitrary[SortedSet[A]] = {
+    // We create an arbitrary Ordering[A] which is either the same as Order[A]
+    // or the reverse of it. This is important because the Ordering in use by
+    // a given SortedSet[A] may not be derived from our Order and
+    // scala.math.Ordering does not imply coherence.
+    val orderingGen: Gen[Ordering[A]] =
+      Gen.oneOf(
+        Order[A].toOrdering,
+        Order[A].toOrdering.reverse
+      )
```

Comment:
These seem reasonable 👍 to get a truly arbitrary ordering, I suppose we could use an arbitrary `A => Int` or similar?

Reply:
Not every `A => Int` is a valid ordering though 😄

Reply:
Oh? 😳 how is that?

Reply:
Yeah, the only two globally valid orderings I think we can make arbitrarily are the canonical one and the reverse of it.

Reply:
> you can use if if you add Cogen[A]

Yes, that's what I was thinking :)

Reply:
Oh...I guess it is....

Reply:
We have to add a Cogen constraint, but that seems fine to me.

Reply:
Updated

Reply:
But now we're not testing the fast path. Which is tricky because
> Ironically Ordering instances created from Order are neither == nor eq 😞

I think you can achieve it if you cache the `Ordering` from `Order` and use it in `Gen.oneOf` like:
```scala
val canonical = Order[A].toOrdering
Gen.oneOf(canonical, canonical.reverse, Arbitrary.arbitrary[A => Int].map(Ordering.by))
```

Reply:
Probably missed some `Gen.const` there but you get the idea

Reply:
@armanbilge I don't think we can use `Arbitrary[A => Int]` after all. Unless I'm missing something, the arbitrary functions generated by Scalacheck do not guarantee that each `A` maps to a `Int` value that is not shared by any other `A`, e.g. we can get a function like this.

```scala
scala> val f = Arbitrary.arbitrary[Short => Int].sample.get
val f: Short => Int = org.scalacheck.GenArities$$Lambda$9321/0x00000008017d9e58@58794be4

scala> f(-32763)
val res7: Int = 1

scala> f(-32764)
val res8: Int = 1
```

Which means they would compare equal by `Ordering`, but not by `.equals`, which breaks the `SortedSet`. I think we need a bijective function here, but we can't get that from Scalacheck unless I'm mistaken.

Am I missing something here?

Reply:
Huh. If the `Ordering` and `.equals` are inconsistent, is that not also a source of unlawfulness? Since the notion of equivalence would affect the size of the set.

Reply:
Wait, the `SortedSet` should use `Ordering.compare`, not `equals`, right?